doc/developer/design/20210311_prometheus_source.md

```diff
@@ -0,0 +1,144 @@
+
+* (Maybe) don't even expose the `CREATE SOURCE` mechanics externally?
+
+# Description
+
+<!--
+    Describe the approach in detail. If there is no clear frontrunner, feel free to list all approaches in alternatives.
+    If applicable, be sure to call out any new testing/validation that will be required
+-->
+
+To import prometheus metrics, a process must regularly scrape an endpoint. Something like the following would be our external interface:
+
+```sql
+CREATE SOURCE metrics_table_name
+PROMETHEUS ENDPOINT 'http://localhost:9090/metrics'
+WITH (
+    metadata_table = "metrics_table_meta",
+    poll_interval = '1 minute'::interval,
+    retain_data = '5 minutes'::interval
+);
+```
+
+This creates two relations that have schemas similar to the following:
+
+```sql
+CREATE TABLE metrics_table_name (
+    metric TEXT,
+    ts TIMESTAMP,
+    labels JSONB,
+    value float8
+);
+
+CREATE TABLE metrics_table_meta (
+    metric TEXT,
+    metric_type TEXT,
+    help TEXT
+);
+```
+
+Once created, the source polls the endpoint once a minute (via the configured the `poll_interval` duration), and retains metric readings for 5 minutes.
+
+
+# Alternatives
+
+<!-- // Similar to the Description section. List of alternative approaches considered, pros/cons or why they were not chosen -->
+
+There are two major things that we could do differently to fulfill the brief, and a series of things that we could do to polish this design. I'll summarize the former under "Alternative approaches" and the latter under "Alternative details".
+
+## Alternative approaches
+
+### Create an internal metrics reporting channel instead (feeding metrics directly)
+
+We could add a mechanism for providing log events to the dataflow crate (or a crate that df and others depend on), where if a piece of code now increments a prometheus counter, it would instead send an event that gets captured by a listener on a channel, and increments the prometheus cuonter, as well as queue the event for updating of internal metrics views.
+
+The major benefit is that we encapsulate our metrics tooling usage more (making it easier to plug in more ways of recording metric values).
+
+The downsides, however, are somewhat big:
+
+* Things need to be structured such that you can get at the logger where you need to count the metric. Some parts where we currently do this make this somewhat difficult (e.g. the pgwire code, where the currently best way to count bytes sent is in encoding).
+
+* Adding metrics gets more difficult. You not only have to figure out how go get your number counted the right way, you also have to figure out how to aggregate it properly in the dataflow world. We worry that this might mean that useful operational metrics don't get added as soon/as quickly as they'd be necessary, making it harder to run the system.
```

Comment:
I view "figuring out how to aggregate the metric properly" as a good thing. Perhaps if there are easy helpers that "just do the right thing", we can make it easy to add new metrics *and* have them aggregated easily?

 I have this strange world view that all metrics should presented as aggregations, particularly when they are intended for human consumption. There's only so much data that can be presented on a display / read by a human and, for most metrics use cases, that means summarizing events. So, that's my verbose way of saying that I think we should require all metrics have proper aggregation semantics.

Reply:
Ah, I cut off some of the reasoning here, maybe need to expand it: Much like the main difficulty with implementing #5825, the way we handle metrics in code right now makes it very hard (possibly impossible with the current stream abstractions) to pass a channel like that down to, say, the code responsible for serializing to the pg wire protocol. Maybe not insurmountable! But definitely a very heavy lift.

The other thing is that in my experience, it's preferable to provide one "reasonable-enough" metrics-providing abstraction, over making them choose between / hook together several options. This design would let us keep providing that reasonable-enough interface via prometheus internally (:

Reply:
I totally agree with having there be a single abstraction! My main concern here is that our chosen abstraction considers aggregations from the start.